src/task/callback.rs

```diff
@@ -0,0 +1,237 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+//! A [`FnOnce`] or [`FnMut`] closure.
+
+use crate::LocalSpawn;
+
+use std::marker::PhantomData;
+
+/// Tasks that is supported by [`Runner`].
+///
+/// Anything that can be converted to a [`Task`] is supported by [`Runner`].
+pub trait CallbackTask<Spawn> {
+    /// Performs the conversion to a callback task.
+    fn as_mut_task(&mut self) -> &mut Task<Spawn>;
+}
+
+impl<Spawn> CallbackTask<Spawn> for Task<Spawn> {
+    fn as_mut_task(&mut self) -> &mut Task<Spawn> {
+        self
+    }
+}
+
+/// A callback task, which is either a [`FnOnce`] or a [`FnMut`].
+pub enum Task<Spawn> {
+    /// A [`FnOnce`] task.
+    Once(Option<Box<dyn FnOnce(&mut Handle<'_, Spawn>) + Send>>),
+    /// A [`FnMut`] task.
+    Mut(Box<dyn FnMut(&mut Handle<'_, Spawn>) + Send>),
+}
+
+impl<Spawn> Task<Spawn> {
+    /// Creates a [`FnOnce`] task.
+    pub fn new_once(t: impl FnOnce(&mut Handle<'_, Spawn>) + Send + 'static) -> Self {
+        Task::Once(Some(Box::new(t)))
+    }
+
+    /// Creates a [`FnMut`] task.
+    pub fn new_mut(t: impl FnMut(&mut Handle<'_, Spawn>) + Send + 'static) -> Self {
+        Task::Mut(Box::new(t))
+    }
+}
+
+/// Handle passed to the task closure.
+///
+/// It can be used to spawn new tasks or control whether this task should be
+/// rerun.
+pub struct Handle<'a, Spawn> {
+    spawn: &'a mut Spawn,
+    rerun: bool,
+}
+
+impl<'a, Spawn> Handle<'a, Spawn>
+where
+    Spawn: LocalSpawn<Task = Task<Spawn>>,
+{
+    /// Spawns a [`FnOnce`] to the thread pool.
```

Comment:
Hmm, should "Fn" pronounce like "F-N", "fun" or "function"? /cc @Hoverbear @nrc @brson

Reply:
I always thought of it as "fun" 🤣I think it's as you please.

Reply:
I say something like "fun" but without the "u" similar to std